src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs

```diff
@@ -3505,5 +3505,320 @@ void M<T>(object e)
+            var model = (CSharpSemanticModel)comp.GetSemanticModel(tree);
+            var binder = model.GetEnclosingBinder(@switch.SpanStart);
+            var boundSwitch = (BoundSwitchStatement)binder.BindStatement(@switch, BindingDiagnosticBag.Discarded);
+            AssertEx.Equal(
+@"[0]: t0 is string ? [1] : [8]
+[1]: t1 = (string)t0; [2]
+[2]: t1 == ""a"" ? [3] : [4]
+[3]: leaf `case ""a"":`
+[4]: t2 = t1.Length; [5]
+[5]: t2 == 1 ? [6] : [13]
+[6]: when <true> ? [7] : <unreachable>
+[7]: leaf `case string { Length: 1 } s:`
+[8]: t0 is int ? [9] : [13]
+[9]: t3 = (int)t0; [10]
+[10]: t3 < 42 ? [11] : [12]
+[11]: leaf `case int and < 42:`
+[12]: when ((i % 2) == 0) ? [14] : [13]
+[13]: leaf `default`
+[14]: leaf `case int i when (i % 2) == 0:`
+", boundSwitch.DecisionDag.Dump());
+        }
+
+        [Fact, WorkItem(53868, "https://github.com/dotnet/roslyn/issues/53868")]
+        public void DecisionDag_Dump_SwitchStatement_02()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    void Deconstruct(out int i1, out string i2, out int? i3)
+    {
+        i1 = 1;
+        i2 = ""a"";
+        i3 = null;
+    }
+
+    void M(C c)
+    {
+        switch (c)
+        {
+            case null:
+                Console.Write(0);
+                break;
+            case (42, ""b"", 43):
+                Console.Write(1);
+                break;
+            case (< 10, { Length: 0 }, { }):
+                Console.Write(2);
+                break;
+            case (< 10, object): // 1, 2
+                Console.Write(3);
+                break;
+        }
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (26,18): error CS7036: There is no argument given that corresponds to the required formal parameter 'i3' of 'C.Deconstruct(out int, out string, out int?)'
+                //             case (< 10, object): // 1, 2
+                Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "(< 10, object)").WithArguments("i3", "C.Deconstruct(out int, out string, out int?)").WithLocation(26, 18),
+                // (26,18): error CS8129: No suitable 'Deconstruct' instance or extension method was found for type 'C', with 2 out parameters and a void return type.
+                //             case (< 10, object): // 1, 2
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "(< 10, object)").WithArguments("C", "2").WithLocation(26, 18)
+            );
+
+            var tree = comp.SyntaxTrees.Single();
+            var @switch = tree.GetRoot().DescendantNodes().OfType<SwitchStatementSyntax>().Single();
+            var model = (CSharpSemanticModel)comp.GetSemanticModel(tree);
+            var binder = model.GetEnclosingBinder(@switch.SpanStart);
+            var boundSwitch = (BoundSwitchStatement)binder.BindStatement(@switch, BindingDiagnosticBag.Discarded);
+            AssertEx.Equal(
+@"[0]: t0 == null ? [1] : [2]
```

Comment:
Another thing I find really useful is the tree dump of the final dag using `rootDecisionDagNode.Dump()` after:

https://github.com/dotnet/roslyn/blob/35ebde268069bacab998ff022104d8eeac05a7e3/src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs#L652

I usually find myself using that because it's easier to reason about and it reflects the final codegen.

However, currently it uses the generic tree dumper with a lot of noise which makes it hard to see what's going on.

Reply:
Interesting, do you think you could mock up a before/after of the dump that contains the information you want? I guess I am wondering if it seems like we should, for example, repeatedly print certain nodes in the dump when there are multiple ways to reach that node. It also feels like we would want to cut out all the "source" properties in this representation since those basically always point "backwards".

Reply:
Something like:
```
testDecisionDagNode
├─test: <debugger display>
├─whenTrue
│ └─evaluationDecisionDagNode
│   ├─evaluation: <debugger display>
│   └─next:
|     └─testDecisionDagNode
|       ├─whenTrue
|       |  └─evaluationDecisionDagNode
|       |    ├─evaluation: <debugger display>
|       |    └─next
|       |      └─testDecisionDagNode
|       |        ├─test: <debugger display>
|       |        ├─whenTrue: leaf
|       |        └─whenFalse
...
```
Instead of:
```
testDecisionDagNode
├─test
│ └─dagNonNullTest
│   ├─isExplicitTest: True
│   ├─input
│   │ └─dagTemp
│   │   ├─type: X
│   │   ├─source
│   │   ├─index: 0
│   │   └─hasErrors: False
│   └─hasErrors: False
├─whenTrue
│ └─evaluationDecisionDagNode
...

Reply:
I think I won't have time to do this myself but would be happy to accept a change which changes the implementation of Dump() on dag nodes to work in this manner.